### PR TITLE
feat(ivy): support ng-content projection in the ivy compiler

### PR DESCRIPTION
**Pull request by chuckjaz**
## PR Type
What kind of change does this PR introduce?

```
[x] Feature
```

## What is the current behavior?

The ivy compiler doesn't support content projection.

## What is the new behavior?

The ivy compiler supports content projection.

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

### CODE REVIEW

File: packages/compiler/src/render3/r3_view_compiler.ts [hunk 1]
```diff
@@ -563,3 +600,66 @@ function invalid<T>(arg: o.Expression | o.Statement | TemplateAst): never {
...
+function getContentProjection(asts: TemplateAst[], ngContentSelectors: string[]) {
+  const projectIndexMap = new Map<NgContentAst, NgContentInfo>();
+  const visitor = new ContentProjectionVisitor(projectIndexMap, ngContentSelectors);
+  templateVisitAll(visitor, asts);
+  return projectIndexMap;
+}
+
+type R3SimpleCssSelector = (string | null)[];
+type R3CssSelectorWithNegations = [R3SimpleCssSelector, null] |
+    [R3SimpleCssSelector, R3SimpleCssSelector] | [null, R3SimpleCssSelector];
```
Comment (vicb) [inline packages/compiler/src/render3/r3_view_compiler.ts]: could positive be `null` ?

File: packages/compiler/src/render3/r3_view_compiler.ts [hunk 2]
```diff
@@ -563,3 +600,66 @@ function invalid<T>(arg: o.Expression | o.Statement | TemplateAst): never {
...
+type R3CssSelector = R3CssSelectorWithNegations[];
+
+function parserSelectorToSimpleSelector(selector: CssSelector): R3SimpleCssSelector {
+  const classes =
+      selector.classNames && selector.classNames.length ? ['class', ...selector.classNames] : [];
```
Comment (vicb) [inline packages/compiler/src/render3/r3_view_compiler.ts]: document the format on the type ?

**Review by vicb — CHANGES_REQUESTED** [review]
a few nits

File: packages/compiler/src/render3/r3_view_compiler.ts (hunk 2 above)
Comment (chuckjaz) [reply packages/compiler/src/render3/r3_view_compiler.ts]: Add a reference to the documentation in core.

Comment (vicb) [reply packages/compiler/src/render3/r3_view_compiler.ts]: it's more about the class part `[..., class, ...classNames]`

**Review by vicb — APPROVED** [review]
one doc nit

File: packages/compiler/src/render3/r3_view_compiler.ts (hunk 2 above)
Comment (chuckjaz) [reply packages/compiler/src/render3/r3_view_compiler.ts]: Yes. That is documented as part of the CssSelector in the referenced file.

**Pull request closed without merging**